### PR TITLE
[#17] Allow underscores in integers

### LINKED ISSUE #17: Add parsing of _ in numbers
TOML allows you to specify numbers like `1_000_000`. This is not supported currently. But probably should  be.

### PR DESCRIPTION
**Pull request by tomphp**
This PR aims to allow underscores in number as mentioned in #17. 

### Some facts about this implementation
- I've used existing commented out tests and assumed they contain the desired behaviour (the TOML spec doesn't cover some of the edge cases in the commented tests).
- This only works for integers.

### Some concerns about the implementation (any suggestions to improve work but appreciated):
- The removing of the underscores is done in a separate action from the parsing. I can use explicit parsers rather than the `match` function, but it's less concise.
- I'm not happy with the use of `undefined` in `fromRight` but the parser should ensure this never happens.

Resolves #17

### CODE REVIEW

File: src/Toml/Parser.hs [hunk 1]
```diff
@@ -159,10 +161,19 @@ tableNameP = lexeme $ between (char '[') (char ']') keyP
 
 -- Values
 
+decimalP :: Parser Integer
+decimalP = mkInteger <$> decimalStringP
+  where
+    decimalStringP   = fst <$> match (some digitChar >> optional (try $ many _digitsP))
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: The usage of `match` looks interesting 🤔It looks like and efficient enough way to get the part of the input. Well done!

File: src/Toml/Parser.hs [hunk 2]
```diff
@@ -159,10 +161,19 @@ tableNameP = lexeme $ between (char '[') (char ']') keyP
...
+    _digitsP         = char '_' >> some digitChar
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: I'm not really sure that `try $ many smth` is an efficient parser. Looks like you can move `try` to `_digitsP`, like:

```haskell
try (char '_') *> some digitChar
```

Also, if you replace `some digitChar` with `many digitChar` you can remove `optional` from `decimalStringP`. I hope it works! And if it works, it will make the code cleaner.

File: src/Toml/Parser.hs [hunk 3]
```diff
@@ -159,10 +161,19 @@ tableNameP = lexeme $ between (char '[') (char ']') keyP
...
+    mkInteger        = textToInt . stripUnderscores
+    textToInt        = fst . fromRight undefined . TR.decimal
+    stripUnderscores = Text.replace "_" ""
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: I think `Text.filter` is more efficient here than `replace`:

* https://hackage.haskell.org/package/text-1.2.3.1/docs/Data-Text.html#v:filter

File: test/Test/Toml/Parsing/Unit.hs [hunk 4]
```diff
@@ -161,15 +162,15 @@ spec_Parser = do
...
-          --  parseInt "13_" 13
-          --  intFailOn "_123_"
-          --  intFailOn "_13"
-          --  intFailOn "_"
+            it "can parse numbers with underscores between digits" $ do
+               parseInteger "1_000" 1000
+               parseInteger "5_349_221" 5349221
+               parseInteger "1_2_3_4_5" 12345
+               parseInteger "1_2_3_" 1
+               parseInteger "13_" 13
```
Comment (ChShersh) [inline test/Test/Toml/Parsing/Unit.hs]: That's great that you found these tests and uncommented them! These tests should be enough.

File: src/Toml/Parser.hs [hunk 5]
```diff
@@ -159,10 +161,19 @@ tableNameP = lexeme $ between (char '[') (char ']') keyP
 
 -- Values
 
+decimalP :: Parser Integer
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: I think it's a good idea for you to open an issue in the `megaparsec` library and ask to add some default parser for integers with underscores. You can link `megaparsec` maintainer to your implementation under this PR as an example of such parser 🙂 I think it's a good idea to have such parser in standard parser-combinators library. Especially in the context that there's support for `_` in numeric literals in Haskell since GHC 8.6.1 (enabled by `NumericUnderscores` language extension). And if such parser appears in `megaparsec`, we can replace our parser in `tomland` with more efficient and standard one. I think this will happen only about one month in `megaparec` library and we still can have this feature in `tomland`. 

I'm okay if the parser in `megaparsec` doesn't parse things like `13_`, it's still better to use standard function, especially then the specification of configuration format is not precise enough.

File: src/Toml/Parser.hs (hunk 2 above)
Comment (tomphp) [reply src/Toml/Parser.hs]: It needs to be this way to pass the test `1_2_3_ == 1`. If the `optional` is removed and the `try` is moved into `_digitsP` then you get `1_2_3_ == 123`. I don't really know what the correct behaviour should be here so happy to change it if the second one is valid?

Comment (ChShersh) [reply src/Toml/Parser.hs]: Oops, that was unfortunate. Yeah, `1_2_3_ == 1` test is invalid, it should be `123` so feel free to change this particular test 👌

File: src/Toml/Parser.hs (hunk 5 above)
Comment (tomphp) [reply src/Toml/Parser.hs]: Great idea. I'll open a PR for megaparsec and come back and update tomland if it's accepted 👍

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,12 @@ Change log
 tomland uses [PVP Versioning][1].
 The change log is available [on GitHub][2].
 
+0.5.1
+=====
+
+* [#97](https://github.com/kowainik/tomland/pull/97)
```
Comment (ChShersh) [inline CHANGELOG.md]: This should go in version 0.5.0 entry because it's not released on Hackage yet.

File: src/Toml/Parser.hs [hunk 7]
```diff
@@ -159,10 +161,19 @@ tableNameP = lexeme $ between (char '[') (char ']') keyP
...
+decimalP = mkInteger <$> decimalStringP
+  where
+    decimalStringP   = fst <$> match (some digitChar >> many _digitsP)
+    _digitsP         = try (char '_' >> some digitChar)
+    mkInteger        = textToInt . stripUnderscores
+    textToInt        = fst . fromRight undefined . TR.decimal
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: `fromRight undefined` is not a pleasant thing... But usually instead of `undefined` it's always better to write `error` with some description, like:

```haskell
fromRight (error "Underscore parser has a bug")
```

File: src/Toml/Parser.hs [hunk 8]
```diff
@@ -159,10 +161,19 @@ tableNameP = lexeme $ between (char '[') (char ']') keyP
 
 -- Values
 
+decimalP :: Parser Integer
+decimalP = mkInteger <$> decimalStringP
+  where
+    decimalStringP   = fst <$> match (some digitChar >> many _digitsP)
+    _digitsP         = try (char '_' >> some digitChar)
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: Shouldn't it be okay to wrap only `char '_'` in `try`? My personal feeling is that `try` can decrease performance a lot, so it's better to reduce scope of `try` parser combinator as much as possible.

Comment (tomphp) [reply src/Toml/Parser.hs]: I tried with it just on the `char '_'` but then I get a parse error on strings that end with an `_` :(

Comment (ChShersh) [reply src/Toml/Parser.hs]: @tomphp From the TOML specification:

> For large numbers, you may use underscores between digits to enhance readability

So it's okay if parser can't parsee number like `13_`. Underscores are allowed only between characters.

Comment (tomphp) [reply src/Toml/Parser.hs]: Done
